Source/WebCore/platform/graphics/nicosia/NicosiaBuffer.cpp

```diff
@@ -143,35 +150,66 @@ void UnacceleratedBuffer::waitUntilPaintingComplete()
 }
 
 #if USE(SKIA)
-Ref<Buffer> AcceleratedBuffer::create(sk_sp<SkSurface>&& surface, Flags flags)
+Ref<Buffer> AcceleratedBuffer::create(Ref<BitmapTexture>&& texture, Flags flags)
 {
-    return adoptRef(*new AcceleratedBuffer(WTFMove(surface), flags));
+    return adoptRef(*new AcceleratedBuffer(WTFMove(texture), flags));
 }
 
-AcceleratedBuffer::AcceleratedBuffer(sk_sp<SkSurface>&& surface, Flags flags)
+AcceleratedBuffer::AcceleratedBuffer(Ref<BitmapTexture>&& texture, Flags flags)
     : Buffer(flags)
+    , m_texture(WTFMove(texture))
 {
-    m_surface = WTFMove(surface);
 }
 
 AcceleratedBuffer::~AcceleratedBuffer()
 {
-    ensureOnMainThread([surface = WTFMove(m_surface), fence = WTFMove(m_fence)]() mutable {
+    // Owned by us, and the BitmapTexturePool from which it is allocated.
+    // If we could potentially destruct the BitmapTexture here, we'd have to
+    // assure this happens on the main thread - but we don't have to.
+    ASSERT(m_texture->refCount() == 2);
+
+    ensureOnMainThread([fence = WTFMove(m_fence)]() mutable {
         PlatformDisplay::sharedDisplay().skiaGLContext()->makeContextCurrent();
         fence = nullptr;
-        surface = nullptr;
     });
 }
 
 WebCore::IntSize AcceleratedBuffer::size() const
 {
-    return { m_surface->width(), m_surface->height() };
+    return m_texture->size();
 }
 
-void AcceleratedBuffer::beginPainting()
+SkCanvas* AcceleratedBuffer::canvas()
 {
-    m_surface->getCanvas()->save();
-    m_surface->getCanvas()->clear(SkColors::kTransparent);
+    if (!PlatformDisplay::sharedDisplay().skiaGLContext()->makeContextCurrent())
+        return nullptr;
+
+    GrGLTextureInfo externalTexture;
+    externalTexture.fTarget = GL_TEXTURE_2D;
+    externalTexture.fID = m_texture->id();
+    externalTexture.fFormat = GL_RGBA8;
+
+    const auto& size = m_texture->size();
+    auto backendTexture = GrBackendTextures::MakeGL(size.width(), size.height(), skgpu::Mipmapped::kNo, externalTexture);
+
+    SkSurfaceProps properties = { 0, FontRenderOptions::singleton().subpixelOrder() };
+    auto surface = SkSurfaces::WrapBackendTexture(PlatformDisplay::sharedDisplay().skiaGrContext(),
+        backendTexture,
+        kTopLeft_GrSurfaceOrigin,
+        PlatformDisplay::sharedDisplay().msaaSampleCount(),
+        kRGBA_8888_SkColorType,
+        SkColorSpace::MakeSRGB(),
+        &properties);
+
+    if (!surface || !surface->getCanvas())
+        return nullptr;
+
+    m_surface = WTFMove(surface);
+
+    auto* canvas = m_surface->getCanvas();
+    canvas->save();
+    canvas->clear(SkColors::kTransparent);
+    return canvas;
```

Comment:
I know I proposed this, but in this patch it's not obvious why we don't do all this in beginPainting(). It's a bit weird that the save and clear is done here and the restore is done in completePainting. Also the name canvas() sounds like a simple getter. I don't remember why we need the save+clear/restore only for accelerated buffers (I guess because we reuse the buffers from a pool), because my initial idea was to move this to the caller, we return the canvas and you should save+clear/restore or whatever. Also note that completePainting is called even when canvas returns nullptr, but getCanvas() is used there without any null check. If we can move this to the caller (assuming it's harmless for CPU), then we can keep this name as canvas() and only create the surface if it doesn't exist.

Reply:
Ok. Will iterate on this.